contracts/nftairdrops.js

```diff
@@ -0,0 +1,461 @@
+};
+
+// END helper functions
+
+// BEGIN contract actions
+
+actions.createSSC = async () => {
+  const tableExists = await api.db.tableExists('pendingAirdrops');
+  if (tableExists === false) {
+    await api.db.createTable('pendingAirdrops', ['airdropId', 'symbol']);
+    await api.db.createTable('params');
+
+    const params = {};
+    params.feePerTransaction = '0.1';
+    params.maxTransactionsPerAirdrop = 50000;
+    params.maxTransactionsPerAccount = 50;
+    params.maxTransactionsPerBlock = 50;
+    params.maxAirdropsPerBlock = 1;
+    params.processingBatchSize = 50;
+    await api.db.insert('params', params);
+  }
+};
+
+actions.updateParams = async (payload) => {
+  if (api.assert(api.sender === api.owner, 'not authorized')) {
+    const {
+      feePerTransaction,
+      maxTransactionsPerAirdrop,
+      maxTransactionsPerAccount,
+      maxTransactionsPerBlock,
+      maxAirdropsPerBlock,
+      processingBatchSize,
+    } = payload;
+
+    const params = await api.db.findOne('params', {});
+
+    if (feePerTransaction !== undefined) {
+      if (api.assert(typeof feePerTransaction === 'string' && !api.BigNumber(feePerTransaction).isNaN() && api.BigNumber(feePerTransaction).gte(0), 'invalid feePerTransaction')) {
+        params.feePerTransaction = feePerTransaction;
+      }
+    }
+    if (maxTransactionsPerAirdrop !== undefined) {
+      if (api.assert(Number.isInteger(maxTransactionsPerAirdrop) && maxTransactionsPerAirdrop > 0, 'invalid maxTransactionsPerAirdrop')) {
+        params.maxTransactionsPerAirdrop = maxTransactionsPerAirdrop;
+      }
+    }
+    if (maxTransactionsPerAccount !== undefined) {
+      if (api.assert(Number.isInteger(maxTransactionsPerAccount) && maxTransactionsPerAccount > 0 && maxTransactionsPerAccount <= params.maxTransactionsPerAirdrop, 'invalid maxTransactionsPerAccount')) {
+        params.maxTransactionsPerAccount = maxTransactionsPerAccount;
+      }
+    }
+    if (maxTransactionsPerBlock !== undefined) {
+      if (api.assert(Number.isInteger(maxTransactionsPerBlock) && maxTransactionsPerBlock > 0, 'invalid maxTransactionsPerBlock')) {
+        params.maxTransactionsPerBlock = maxTransactionsPerBlock;
+      }
+    }
+    if (maxAirdropsPerBlock !== undefined) {
+      if (api.assert(Number.isInteger(maxAirdropsPerBlock) && maxAirdropsPerBlock > 0, 'invalid maxAirdropsPerBlock')) {
+        params.maxAirdropsPerBlock = maxAirdropsPerBlock;
+      }
+    }
+    if (processingBatchSize !== undefined) {
+      if (api.assert(Number.isInteger(processingBatchSize) && processingBatchSize > 0, 'invalid processingBatchSize')) {
+        params.processingBatchSize = processingBatchSize;
+      }
+    }
+
+    await api.db.update('params', params);
+
+    delete params._id; /* eslint-disable-line no-underscore-dangle */
+    api.emit('paramsUpdated', params);
+  }
+};
+
+actions.newAirdrop = async (payload) => {
+  const {
+    symbol,
+    list,
+    startBlockNumber,
+    softFail,
+    fromType,
```

Comment:
As you indicated the NFT contract currently doesn't support using fromType as 'contract', it would be good to control whether or not this feature is enabled with a boolean params config field.  For now we can leave it disabled (meaning it would be an input validation error if the user sets this to 'contract'), then if the NFT contract supports it in the future, we can turn on support by updating the params.

Reply:
Added extra setting in params-table to be able to control the allowed fromType.